drivers: regulator: more improvements

gmarull opened the pull request:
- Minor cleanups to the regulator code
- Add new `regulator_mode_get` API
- Add support for `regulator-initial-mode`
- Refactor shell

gmarull committed:
drivers: regulator: common: remove redundant cast

Unlike in C++ code, explicit casts from void* are not needed, so remove
them.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
tests: drivers: regulator: api: check init conditions during setup

Check conditions that happened during device init at setup stage, so
that we can reset fakes and make tests stateless.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: regulator: apply initial mode

Apply initial mode (regulator-initial-mode) as part of the
regulator_common_init_enable call. Update tests to cover this change.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: regulator: add get_mode API

Add a new API to query the configured regulator mode. Updated fake
driver and API tests.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: regulator: npm6001: add support for get_mode

Allow to query operating mode on each BUCK converter (PWM or
Hysteretic).

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
samples: shields: npm6001_ek: add support to query mode

Add a new command to query the regulator operating mode.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
samples: shields: npm6001_ek: remove dead assignment

The voltage gets assigned by regulator_get_voltage, code was likely a
copy&paste from set voltage command.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: regulator: shell: refactor shell

- Refactor the regulator shell so that it exposes all regulator APIs
- vset/iset commands allow to specify a single value (equal min/max) or
  a range
- Voltage/current input is now more user friendly, e.g. user can specify
  units and decimals: 3.3v, 200mv, -4mv, etc.
- Reported values are also printed in a more user friendly way, e.g.
  1800000 uV will be printed as 1.800 V.
- Added new command to list supported voltages

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
maintainers: add entry for regulator drivers

Add entry for regulator drivers. Add myself as a maintainer as I've been
reworking the whole subsystem lately, and Daniel DeGrasse as a
collaborator as he's been active in reviews and implementations.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

danieldegrasse commented on drivers/regulator/regulator_shell.c:
```diff
@@ -1,152 +1,404 @@
 	return 0;
 }
 
+static int cmd_vlist(const struct shell *sh, size_t argc, char **argv)
```
I wanted to get your opinion on an issue I had while using this command with the PCA9420. Since this regulator supports discontinuous voltage ranges (and therefore uses the `liner_range` API), `regulator_count_voltage` will return the number of supported voltage levels (as expected). However, on the PCA9420, this is different from the number of valid voltage indices usable in `regulator_list_voltage`. The result is that not all supported voltages will be listed using this command, since the number of supported voltages is less than the number of indices available for `regulator_list_voltage`

There are a few ways I can think of to resolve this:
- directly record the number of supported voltages for each regulator in the PCA9420 driver, and return this value for an API call to `regulator_count_voltages`
- Modify the `linear_range` API so that `linear_range_group_values_count` counts the number of indices that are valid for a given linear range.

Of the two options, modifying the `linear_range` API makes more sense to me. I'm not sure what cases exist where a user would only want the number of discrete supported values, and not the number of valid indices (since they will probably iterate through the indices if they are using this API)

gmarull replied:
This is actually an oversight that needs to be fixed. I think the fix needs to go into the linear_range API as you suggest. Let me fix it.

gmarull replied:
Fixed the linear_range API. I also adjusted shell so that it doesn't print repeated values. Thanks for reporting this.

danieldegrasse commented on drivers/regulator/regulator_shell.c:
```diff
@@ -1,152 +1,404 @@
+		      "Usage: vlist <device>",
+		      cmd_vlist, 2, 0),
+	SHELL_CMD_ARG(vset, NULL,
+		      "Set voltage\n"
```
The parsing of volts and mV implemented in this command is very useful, but I think we should document it in the help string (maybe by clarifying the user should provide the unit with their input?)

gmarull replied:
Sure, let me add a comment.

danieldegrasse reviewed (commented):
Overall, looks good- just a few minor comments

gmarull committed:
lib: linear_range: fix count for ranges with step 0

Ranges with step 0, that is, multiple indices representing the same
value, need to report the count exclusively based on indices. This makes
it possible to later retrieve all values, e.g. in a group.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>
